okonomiyaki/versions/semver.py

```diff
@@ -0,0 +1,197 @@
+
+
+def _ensure_no_leading_zero(value, name):
+    if len(value) > 1 and value.startswith("0"):
+        msg = "{0} number cannot have leading 0: {1!r}".format(name, value)
+        raise ValueError(msg)
+
+
+def _parse_pre_release(s):
+    if s is not None:
+        return tuple(part for part in s[1:].split("."))
+    else:
+        return None
+
+
+def _parse_build(s):
+    return _parse_pre_release(s)
+
+
+def _convert_pre_release(part):
+    try:
+        value = int(part)
+    except ValueError:
+        return part
+    else:
+        _ensure_no_leading_zero(part, "Pre release part")
+        return value
+
+
+class _PrereleaseParts(object):
+    """ Private class used to compare the pre release and build parts. We need
+    this as an empty tuple need to compare greated than any non empty tuple.
+    """
+    def __init__(self, parts):
+        self._comparable_parts = tuple(_convert_pre_release(p) for p in parts)
+
+    def _compare_parts(self, left_parts, right_parts):
+        for left, right in zip(left_parts, right_parts):
+            if left == right:
+                continue
+            else:
+                is_left_int = isinstance(left, (long, int))
+                is_right_int = isinstance(right, (long, int))
+                if is_left_int:
+                    if is_right_int:
+                        return left < right
+                    else:
+                        return True
+                else:
+                    if is_right_int:
+                        return False
+                    else:
+                        return left < right
+        return len(left_parts) < len(right_parts)
+
+    def __hash__(self):
+        return hash(self._comparable_parts)
+
+    def __eq__(self, other):
+        assert isinstance(other, self.__class__)
+        return self._comparable_parts == other._comparable_parts
+
+    def __ne__(self, other):
+        return not (self == other)
+
+    def __lt__(self, other):
+        assert isinstance(other, self.__class__)
+        if self._comparable_parts == other._comparable_parts:
+            return False
+        elif len(self._comparable_parts) == 0:
+            return False
+        elif len(other._comparable_parts) == 0:
+            return True
+        else:
+            return self._compare_parts(
+                self._comparable_parts, other._comparable_parts
+            )
+
+    def __le__(self, other):
+        return self == other or self < other
+
+    def __gt__(self, other):
+        return not (self <= other)
+
+    def __ge__(self, other):
+        return not (self < other)
+
+
+class SemanticVersion(object):
+    """ 'Semver' 2.0 implementation.
+
+    This class takes care of parsing and comparing semver objects.
+    """
+    @classmethod
+    def from_string(cls, s):
+        m = _SEMVER_R.match(s)
+
+        if m is None:
+            raise ValueError("Invalid semver: {0!r}".format(s))
+        else:
+            major = m.group(1)
```

Comment:
I think it would be better to use named groups and access with `m.group('major')`. Reading this, I wonder if the first group should be `0` or `1`; given the tests this is probably right, but it is inconsistent with the usual index-based accessing and can lead to confusion.